torchdata/datapipes/iter/util/_archivereader.py

```diff
@@ -0,0 +1,29 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+from collections import defaultdict
+from io import BufferedIOBase
+from typing import DefaultDict, List, Tuple, Union, IO
+
+from torch.utils.data import IterDataPipe
+
+
+class _ArchiveReaderIterDataPipe(IterDataPipe[Tuple[str, BufferedIOBase]]):
+    def __init__(self):
+        super().__init__()
+        self.open_source_streams: DefaultDict[str, List[Union[BufferedIOBase, IO[bytes]]]] = defaultdict(list)
+        self.open_archive_streams: DefaultDict[str, List[Union[BufferedIOBase, IO[bytes]]]] = defaultdict(list)
```

Comment:
This has another potential issue that `open_streams` are appended inside each iterator. What if two iterators are constructed, there would be duplicated filename. If I only want to close the opened streams inside one iterator, all the streams are closed.

Reply:
Is there any way for us to check if the stream has been depleted?

Reply:
I can't think of a good way without trying call `read()` or something similar directly. There is also the possibility of the user rewinding the stream if they wanted to.

Therefore, I prefer to give users the ability to close the stream rather than detecting if the stream has ended and close it for them, especially true for `FileLoader` streams that is used by multiple DataPipes.

Reply:
Could we rely on https://docs.python.org/3/library/gc.html#gc.callbacks?

Reply:
gc should know the reference for the opened file handle is out of scope. The warning came from gc when the file is opened and not used.
If we have a way to register the opened file handle to gc to run close for us, we may be able to suppress the warning.

But, we need to make sure the file handle registered to gc don't increment the ref_count. Otherwise, the opened file handle would never be closed even with the callback.